pkg/reconciler/pipelinerun/pipelinerun.go

```diff
@@ -1183,34 +1198,107 @@ func (c *Reconciler) updatePipelineRunStatusFromInformer(ctx context.Context, pr
 		logger.Errorf("could not list TaskRuns %#v", err)
 		return err
 	}
-	updatePipelineRunStatusFromTaskRuns(logger, pr, taskRuns)
-
 	runs, err := c.runLister.Runs(pr.Namespace).List(k8slabels.SelectorFromSet(pipelineRunLabels))
 	if err != nil {
 		logger.Errorf("could not list Runs %#v", err)
 		return err
 	}
-	updatePipelineRunStatusFromRuns(logger, pr, runs)
+
+	updatePipelineRunStatusFromChildObjects(ctx, logger, pr, taskRuns, runs)
+
+	return validateChildObjectsInPipelineRunStatus(ctx, pr.Status)
+}
+
+func updatePipelineRunStatusFromChildObjects(ctx context.Context, logger *zap.SugaredLogger, pr *v1beta1.PipelineRun, taskRuns []*v1beta1.TaskRun, runs []*v1alpha1.Run) {
+	taskNameByPipelineTask := make(map[string]string)
+
+	cfg := config.FromContextOrDefaults(ctx)
+	fullEmbedded := cfg.FeatureFlags.EmbeddedStatus == config.FullEmbeddedStatus || cfg.FeatureFlags.EmbeddedStatus == config.BothEmbeddedStatus
+	minimalEmbedded := cfg.FeatureFlags.EmbeddedStatus == config.MinimalEmbeddedStatus || cfg.FeatureFlags.EmbeddedStatus == config.BothEmbeddedStatus
+
+	if minimalEmbedded {
+		taskNameByPipelineTask = updatePipelineRunStatusFromChildRefs(logger, pr, taskRuns, runs)
+	}
+	if fullEmbedded {
+		updatePipelineRunStatusFromTaskRuns(logger, pr, taskRuns, taskNameByPipelineTask)
+		updatePipelineRunStatusFromRuns(logger, pr, runs)
+	}
+}
+
+func validateChildObjectsInPipelineRunStatus(ctx context.Context, prs v1beta1.PipelineRunStatus) error {
+	cfg := config.FromContextOrDefaults(ctx)
+
+	// Verify that we don't populate child references for "full"
+	if cfg.FeatureFlags.EmbeddedStatus == config.FullEmbeddedStatus && len(prs.ChildReferences) > 0 {
+		return fmt.Errorf("expected no ChildReferences with embedded-status=full, but found %d", len(prs.ChildReferences))
+	}
+
+	// Verify that we don't populate TaskRun statuses for "minimal"
+	if cfg.FeatureFlags.EmbeddedStatus == config.MinimalEmbeddedStatus && len(prs.TaskRuns) > 0 {
+		return fmt.Errorf("expected no TaskRun statuses with embedded-status=minimal, but found %d", len(prs.TaskRuns))
+	}
+	// Verify that we don't populate Run statuses for "minimal"
+	if cfg.FeatureFlags.EmbeddedStatus == config.MinimalEmbeddedStatus && len(prs.Runs) > 0 {
+		return fmt.Errorf("expected no Run statuses with embedded-status=minimal, but found %d", len(prs.Runs))
+	}
+
+	// Verify that the TaskRun and Run statuses match the ChildReferences for "both"
+	if cfg.FeatureFlags.EmbeddedStatus == config.BothEmbeddedStatus {
+		if len(prs.ChildReferences) != len(prs.TaskRuns)+len(prs.Runs) {
+			return fmt.Errorf("expected the same number of ChildReferences as the number of combined TaskRun and Run statuses (%d), but found %d",
+				len(prs.TaskRuns)+len(prs.Runs),
+				len(prs.ChildReferences))
+		}
+
+		for _, cr := range prs.ChildReferences {
```

Comment:
thanks for adding this function! just a few quick comments:
- I would collect the names of child refs not in taskruns/runs and return an error containing all of them; this is more informative
- make sure you're handling the case where childRef isn't a taskrun or a run
- please add some tests for this function

Reply:
👍 on all of that!